### ISSUE TITLE
Rule Change: prefer-object-has-own and hasOwnProperty

### ISSUE THREAD

**Issue by erosman** [eslint/eslint#17765]
### What rule do you want to change?

prefer-object-has-own 

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
/*eslint no-prototype-builtins: "error"*/
/*eslint prefer-object-has-own: "error"*/

const obj = {a: 1, b: 2};

obj.hasOwnProperty('a');
Object.hasOwnProperty.call(obj, 'a');
Object.prototype.hasOwnProperty(obj, 'a');
Object.prototype.hasOwnProperty.call(obj, 'a');
```

### What does the rule currently do for this code?

IMHO, the rule should also mark `obj.hasOwnProperty('a')`.

### What will the rule do after it's changed?

Report  `obj.hasOwnProperty()`

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

Previously: #17751

**Comment by mdjermanovic** [comment]
> IMHO, the rule should also mark `obj.hasOwnProperty('a')`.

`no-prototype-builtins` rule already reports this code as a possible logic error, for the reasons described in the [documentation](https://eslint.org/docs/latest/rules/no-prototype-builtins).

`prefer-object-has-own` is a stylistic rule that enforces the use of `Object.hasOwn()` over equivalent alternatives like `Object.prototype.hasOwnProperty.call()`, so I don't think it should report `obj.hasOwnProperty('a')` as it's not equivalent and calling `hasOwnProperty` directly on an object is not just a stylistic issue.

**Comment by erosman** [comment]
> [Object.prototype.hasOwnProperty()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty)
> 
> **Note:** [Object.hasOwn()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn) is recommended over `hasOwnProperty()`, in browsers where it is supported.
> 
> ```jd
> const object1 = {};
> object1.property1 = 42;
> 
> console.log(object1.hasOwnProperty('property1'));
> // Expected output: true
> 
> console.log(object1.hasOwnProperty('toString'));
> // Expected output: false
> 
> console.log(object1.hasOwnProperty('hasOwnProperty'));
> // Expected output: false
> ```

Conceptually, `prefer-object-has-own` comes across as a rule to recommend `Object.hasOwn()` over `hasOwnProperty()`. Limiting it to a specific format seems to be less inclusive.

```js
/*eslint prefer-object-has-own: "error"*/

const obj = {a: 1, b: 2};

obj.hasOwnProperty('a');
Object.prototype.hasOwnProperty(obj, 'a');

Object.hasOwnProperty.call(obj, 'a');
Object.prototype.hasOwnProperty.call(obj, 'a');
```

### Footnote
This is only  a suggestion. In fact as mentioned in #17751, `prefer-object-has-own`  is a subset of `no-prototype-builtins` and as a result, I have removed it from the config that I use.

**Comment by mdjermanovic** [comment]
It's not a subset. Code reported by `prefer-object-has-own` is not reported by `no-prototype-builtins`. Here's a [Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgcHJlZmVyLW9iamVjdC1oYXMtb3duOiBcImVycm9yXCIqL1xuLyplc2xpbnQgbm8tcHJvdG90eXBlLWJ1aWx0aW5zOiBcImVycm9yXCIqL1xuXG5PYmplY3QucHJvdG90eXBlLmhhc093blByb3BlcnR5LmNhbGwob2JqLCBcImFcIik7XG5cbk9iamVjdC5oYXNPd25Qcm9wZXJ0eS5jYWxsKG9iaiwgXCJhXCIpO1xuXG4oe30pLmhhc093blByb3BlcnR5LmNhbGwob2JqLCBcImFcIik7XG5cbmNvbnN0IGhhc1Byb3BlcnR5ID0gT2JqZWN0LnByb3RvdHlwZS5oYXNPd25Qcm9wZXJ0eS5jYWxsKG9iamVjdCwgcHJvcGVydHkpOyIsIm9wdGlvbnMiOnsiZW52Ijp7fSwicnVsZXMiOnt9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSJ9fX0=) showing this.

**Comment by erosman** [comment]
I see..  I misunderstood.

I thought both the `Object.prototype.hasOwnProperty()` and `Object.prototype.hasOwnProperty.call()` would come under `no-prototype-builtins`.

**Comment by mdjermanovic** [comment]
> I thought both the `Object.prototype.hasOwnProperty()` and the `Object.prototype.hasOwnProperty.call()` would come under `no-prototype-builtins`.

`no-prototype-builtins` doesn't disallow the use of `Object.prototype.hasOwnProperty`. This rule just enforces calling it by `Object.prototype.hasOwnProperty.call()` or similar methods over calling it through objects.

I'm closing this issue since the question about the purpose of these two rules seems to be resolved.
